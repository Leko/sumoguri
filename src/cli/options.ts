import Yargs from 'yargs'
import { Viewport } from '../Viewport'

export type CLIOptions = {
  _: string[]
  outDir: string
  parallel: number
  flat: boolean
  include: string[]
  exclude: string[]
  depth: number
  viewports: Viewport[]
  locales: string[]
  prefersColorSchemes: string[]
  disableCssAnimation: boolean
}

export const cmd = Yargs.scriptName('sumoguri')
  .usage('Usage: $0 <entry_url> [options]')
  .example('$0 https://example.com', 'Run with default settings')
  .example(
    '$0 https://example.com --viewports 1680x1050,828x1792',
    'Test in multiple viewports'
  )
  .example(
    '$0 https://example.com --locales en-US,ja-JP',
    'Test in multiple locales'
  )
  .example(
    '$0 https://example.com --prefers-color-schemes=light,dark',
    'Test in multiple color schemes'
  )
  .example('$0 https://example.com --depth 0', 'Do not traverse links')
  .epilog('repository: https://github.com/Leko/sumoguri')
  .options({
    outDir: {
      alias: 'o',
      description: 'Output directory.',
      default: '__screenshots__',
    },
    flat: {
      alias: 'f',
      type: 'boolean',
      description: 'Flatten output file structure.',
      default: false,
    },
    depth: {
      alias: 'd',
      description: 'Depth of traversing.',
      default: 100,
      type: 'number',
    },
    viewports: {
      alias: 'V',
      description: 'Viewports.',
      default: '800x600',
      coerce: (args: string) => {
        return args
          .split(',')
          .map((arg) => arg.split('x').map((n) => parseInt(n, 10)))
      },
    },
    locales: {
      alias: 'l',
      description: 'Locales.',
      default: 'en-US',
      coerce: (args: string) => {
        return args.split(',').map((arg) => arg.trim())
      },
    },
    'prefers-color-schemes': {
      description: 'Specify prefers-color-scheme.',
      default: 'light',
      coerce: (args: string) => {
        return args.split(',').map((arg) => arg.trim())
      },
    },
    'disable-css-animation': {
      description: 'Disable CSS animation and transition.',
      type: 'boolean',
      default: true,
    },
  })

export const parse = cmd.parse.bind(cmd)
