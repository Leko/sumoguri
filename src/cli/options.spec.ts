import { cmd } from './options'

describe('CLI options', () => {
  beforeEach(() => {
    console.error = jest.fn().mockImplementation()
  })
  afterEach(() => {
    ;(console.error as jest.Mock).mockRestore()
  })

  it('help snapshots', () => {
    cmd.showHelp()
    expect((console.error as jest.Mock).mock.calls[0][0]).toMatchSnapshot()
  })
})
