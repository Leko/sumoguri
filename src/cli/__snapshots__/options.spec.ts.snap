// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI options help snapshots 1`] = `
"Usage: sumoguri <entry_url> [options]

Options:
      --help                   Show help                               [boolean]
      --version                Show version number                     [boolean]
  -o, --outDir                 Output directory.    [default: \\"__screenshots__\\"]
  -f, --flat                   Flatten output file structure.
                                                      [boolean] [default: false]
  -d, --depth                  Depth of traversing.      [number] [default: 100]
  -V, --viewports              Viewports.                   [default: \\"800x600\\"]
  -l, --locales                Locales.                       [default: \\"en-US\\"]
      --prefers-color-schemes  Specify prefers-color-scheme.  [default: \\"light\\"]
      --disable-css-animation  Disable CSS animation and transition.
                                                       [boolean] [default: true]

Examples:
  sumoguri https://example.com              Run with default settings
  sumoguri https://example.com --viewports  Test in multiple viewports
  1680x1050,828x1792
  sumoguri https://example.com --locales    Test in multiple locales
  en-US,ja-JP
  sumoguri https://example.com              Test in multiple color schemes
  --prefers-color-schemes=light,dark
  sumoguri https://example.com --depth 0    Do not traverse links

repository: https://github.com/Leko/sumoguri"
`;
